### PR TITLE
[OW] - Refactored code and implemented TODOs from readme

### PR DESCRIPTION
**Pull request by LucianTuduce**
- [x] Rounding when calculating the amount is not done correctly for this type of operation.
- [x] Investigate whether it is possible to implement the value-object serialization, to avoid value nested field in JSON. See #10 as a starting point. Or maybe there is a better solution to the problem at hand?
- [x] Move parameter-specific logic outside the controller.
- [x] Better error handling, especially of potential errors from NBP API.
- [x] Caching the NBP API results.
- [x] Circuit-breaker for the NBP API client.
- [x] Better logging with traceability.
- [x] Replace exceptions with Result (either) which improves the overall methods API readability and forces error handling. Look into cleankod/architecture-archetype as a starting point.
- [x] Test coverage report.
- [x] Auto generating REST API docs. (http://localhost:8080/swagger-ui.html)
- [x] Integration tests with the real NBP API.

### CODE REVIEW

File: src/main/java/pl/cleankod/exchange/core/repository/AccountInMemoryRepositoryImpl.java [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
 
 import pl.cleankod.exchange.core.domain.Account;
 import pl.cleankod.exchange.core.domain.Money;
-import pl.cleankod.exchange.core.gateway.AccountRepository;
 
 import java.util.Optional;
 import java.util.Set;
 
-public class AccountInMemoryRepository implements AccountRepository {
+public class AccountInMemoryRepositoryImpl implements AccountRepository {
```
Comment (steinwelberg) [inline src/main/java/pl/cleankod/exchange/core/repository/AccountInMemoryRepositoryImpl.java:9]: Why did you place the repository in the `core` package?

File: src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java [hunk 2]
```diff
@@ -1,26 +0,0 @@
-package pl.cleankod.exchange.provider;
```
Comment (steinwelberg) [inline src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:1]: Why did you remove the `provider` package?

File: src/main/java/pl/cleankod/util/DateUtils.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class DateUtils {
+
+    private final static String DATE_FORMAT_YYYY_MM_DD = "yyyy-MM-dd";
+
+    public static String getTodayDate() {
+        return LocalDate.now().format(DateTimeFormatter.ofPattern(DATE_FORMAT_YYYY_MM_DD));
```
Comment (steinwelberg) [inline src/main/java/pl/cleankod/util/DateUtils.java:11]: Why do you use the `LocalDate` class?

File: src/main/java/pl/cleankod/exchange/exception/handler/RestExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import pl.cleankod.exchange.exception.error.ErrorMessage;
+
+import static org.springframework.http.HttpStatus.*;
+import static pl.cleankod.exchange.exception.error.ErrorMessageBuilder.buildExceptionMessage;
+
+@Order(Ordered.HIGHEST_PRECEDENCE)
+@RestControllerAdvice
+public class RestExceptionHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(AccountAggregatorImpl.class);
```
Comment (steinwelberg) [inline src/main/java/pl/cleankod/exchange/exception/handler/RestExceptionHandler.java:24]: Why don't you use `Slf4j`?

File: src/test/groovy/pl/cleankod/exchange/AccountSpecification.groovy [hunk 5]
```diff
@@ -1,48 +1,25 @@
 package pl.cleankod.exchange
```
Comment (steinwelberg) [inline src/test/groovy/pl/cleankod/exchange/AccountSpecification.groovy:1]: You forgot to update the package 😉

File: src/test/groovy/pl/cleankod/exchange/AccountSpecification.groovy [hunk 6]
```diff
@@ -1,48 +1,25 @@
...
-import com.github.tomakehurst.wiremock.core.WireMockConfiguration
 import org.apache.http.HttpResponse
 import pl.cleankod.BaseApplicationSpecification
-import pl.cleankod.exchange.core.domain.Account
-import pl.cleankod.exchange.core.domain.Money
+import pl.cleankod.exchange.core.dto.AccountDto
 
 import java.nio.charset.StandardCharsets
 
 class AccountSpecification extends BaseApplicationSpecification {
```
Comment (steinwelberg) [inline src/test/groovy/pl/cleankod/exchange/AccountSpecification.groovy:9]: You removed the unit test for the integration test, why didn't you add the integration test instead?

File: src/main/java/pl/cleankod/exchange/core/domain/Money.java [hunk 7]
```diff
@@ -19,8 +18,5 @@ public static Money of(String amount, String currency) {
         Preconditions.requireNonNull(currency);
         return new Money(new BigDecimal(amount), Currency.getInstance(currency));
     }
-
-    public Money convert(CurrencyConversionService currencyConverter, Currency targetCurrency) {
-        return currencyConverter.convert(this, targetCurrency);
-    }
+
```
Comment (steinwelberg) [inline src/main/java/pl/cleankod/exchange/core/domain/Money.java:21]: Why did you remove the convert method from the `Money` class?

**Review by steinwelberg — CHANGES_REQUESTED** [review]

File: src/test/resources/application.properties [hunk 8]
```diff
@@ -1,2 +1,2 @@
-provider.nbp-api.base-url=http://localhost:8081
-app.base-currency=PLN
\ No newline at end of file
+provider.nbp-api.base-url=http://api.nbp.pl/api
```
Comment (cleankod) [inline src/test/resources/application.properties:1]: If the values are the same as the one in `src/main/resources/application.properties`, then there is no point of overriding.

File: src/main/java/pl/cleankod/exchange/core/cache/CacheServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    public void remove(final String key) {
+        cache.remove(key);
+    }
+
+    @Override
+    public <T> T get(final String key) {
+        return Optional.ofNullable(cache.get(key))
+                .map(SoftReference::get)
+                .filter(cacheObject -> !cacheObject.isExpired())
+                .map(val -> (T) val.value()).orElse(null);
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/cache/CacheServiceImpl.java:36]: This casting could have been avoided if the generics were implemented differently. Can you suggest an improvement?

Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/cache/CacheServiceImpl.java:36]: Maybe return `Optional` instead of `null` here?

File: src/main/java/pl/cleankod/exchange/core/cache/CacheServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+
+    @Override
+    public void remove(final String key) {
+        cache.remove(key);
+    }
+
+    @Override
+    public <T> T get(final String key) {
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/cache/CacheServiceImpl.java:32]: Where and how are the entries acquired from the cache?

File: src/main/java/pl/cleankod/exchange/core/mapper/AccountMapper.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package pl.cleankod.exchange.core.mapper;
+
+import pl.cleankod.exchange.core.domain.Account;
+import pl.cleankod.exchange.core.domain.Money;
+import pl.cleankod.exchange.core.dto.AccountDto;
+import pl.cleankod.exchange.core.dto.MoneyDto;
+
+public class AccountMapper {
+
+    public AccountDto toDto(final Account entity, final Money money) {
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/mapper/AccountMapper.java:10]: What if the DTOs could create themselves from the domain objects? Then we won't need additional classes to do so. What do you think?

File: src/main/java/pl/cleankod/exchange/core/repository/AccountInMemoryRepositoryImpl.java [hunk 12]
```diff
@@ -23,14 +22,14 @@ public class AccountInMemoryRepository implements AccountRepository {
     );
 
     @Override
-    public Optional<Account> find(Account.Id id) {
+    public Optional<Account> find(final Account.Id id) {
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/repository/AccountInMemoryRepositoryImpl.java:25]: What's the purpose of the `final` keyword here?

File: src/main/java/pl/cleankod/exchange/core/service/AccountService.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package pl.cleankod.exchange.core.service;
+
+import pl.cleankod.exchange.core.domain.Account;
+
+public interface AccountService {
+
+    Account execute(final Account.Id id);
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/service/AccountService.java:7]: I don't understand what this service is supposed to do by looking at the class and methods names. How can we improve this?

File: src/main/java/pl/cleankod/exchange/core/service/AccountService.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package pl.cleankod.exchange.core.service;
+
+import pl.cleankod.exchange.core.domain.Account;
+
+public interface AccountService {
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/service/AccountService.java:5]: What is the point of extracting an interface that'd have only one implementation, and putting it in the same package?

File: src/main/java/pl/cleankod/exchange/core/service/AccountServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+    public AccountServiceImpl(final AccountRepository accountRepository) {
+        this.accountRepository = accountRepository;
+    }
+
+    @Override
+    public Account execute(final Account.Id id) {
+        return accountRepository.find(id)
+                .orElseThrow(() -> {
+                    LOGGER.error("Account(id={}) not found", id.value());
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/service/AccountServiceImpl.java:25]: Is that really an error?

File: src/main/java/pl/cleankod/exchange/core/service/AccountServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+                    return new ResourceNotFoundException(AccountErrorType.NOT_FOUND,
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/service/AccountServiceImpl.java:26]: Is there a better way to control the flow, other than throwing exceptions?

File: src/main/java/pl/cleankod/exchange/core/service/CurrencyConversionNbpServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        if (!baseCurrency.equals(targetCurrency)) {
+            LOGGER.info("convert(): Calling currency conversion provider with arguments: table={}, currency={}, date={}",
+                    TABLE_A, targetCurrency.getCurrencyCode(), getTodayDate());
+            final RateWrapper rateWrapper = getRateWrapperAndAddToCache(targetCurrency);
+            final List<Rate> rates = rateWrapper.rates();
+            checkIfRatesArePresent(CollectionUtils.isEmpty(rates));
+            final BigDecimal midRate = rates.get(0).mid();
+            final BigDecimal calculatedRate = money.amount()
+                    .divide(midRate, RoundingMode.HALF_UP)
+                    .setScale(ROUNDING_SCALE, RoundingMode.HALF_UP);
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/service/CurrencyConversionNbpServiceImpl.java:63]: This is not precise enough for money operations.

File: src/main/java/pl/cleankod/exchange/core/service/CurrencyConversionNbpServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+            return new Money(calculatedRate, targetCurrency);
+        }
+
+        if (!money.currency().equals(targetCurrency)) {
+            LOGGER.error("convert(): Failed converting currency {} to {}", money.currency(), targetCurrency);
```
Comment (cleankod) [inline src/main/java/pl/cleankod/exchange/core/service/CurrencyConversionNbpServiceImpl.java:68]: Is that really an error?

**Review by cleankod — CHANGES_REQUESTED** [review]
I only had a chance to review half of the code so my remarks reflect that.
